#!/usr/bin/env python
PACKAGE = "reconfros_software"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

gen.add(                   "use_hsv",   bool_t, 0,                                         "Using HSV color space to unset plants or grass pixels",            False)
gen.add(             "h_upper_bound",    int_t, 0,                                                     "Upper Bound for H Cchannel (only with HSV)",   40,   0,  180)
gen.add(             "h_lower_bound",    int_t, 0,                                                     "Lower Bound for H channel (only with HSV)",   32,   0,  180)
gen.add(                       "dry",   bool_t, 0,                                   "Is the grass dried out near the trail (brown in the image)?",            True)
gen.add(             "difference_gb",    int_t, 0, "Lower bound for the difference of the green and blue channel for considering pixels as plants",   30,   0,  255)
gen.add(         "gauss_kernel_size",    int_t, 0,                                "Number of rows and cols for the gauss convolution (+1 if even)",   13,   1,   50)
gen.add(               "gray_thresh",    int_t, 0,                                                                  "Threshold for the gray value",  150,   0,  255)
gen.add(              "use_contours",   bool_t, 0,                                          "Using findContours() from OpenCV for path extraction",            True)
gen.add(             "use_x_average",   bool_t, 0,                                "Using average x-coordinates of trail-pixels for path extraction",            True)
gen.add(            "avg_block_size",    int_t, 0,                                           "Block size for the average direction control method",   64,   1,  100)
gen.add(             "use_pixel_cnt",   bool_t, 0,                                            "Using the count of trail-pixels for path extraction",            True)
gen.add(     "contour_target_height", double_t, 0,                      "The height of the target-line for direction control (in % of image size)",  0.6, 0.0,  1.0)
gen.add(      "contour_target_range", double_t, 0,               "The search radius on the target-line for direction control (in % of image size)",  0.3, 0.0,  1.0)
gen.add("number_of_horizontal_lines",    int_t, 0,  "Number of horizontal lines where the trail path should be determined (only without contours)",   10,   1, 1000)
gen.add(                 "use_morph",   bool_t, 0,   "Using dilation and erosion to fill holes in the trail and remove fragments from the borders",            True)
gen.add(         "morph_kernel_size",    int_t, 0,                      "Number of rows and cols for the erosion and dilation kernel (+1 if even)",   41,   1,  100)

exit(gen.generate(PACKAGE, "cv_node", "CV"))